### PR TITLE
Enable hipBLASLt 

### PR DESCRIPTION
**Pull request by ahsan-ca**
This PR adds functionality to enable hipBLASLt in MIGraphX.
hipBLASLt can be enabled via MIGRAPHX_ENABLE_HIPBLASLT_GEMM env variable.

### CODE REVIEW

File: src/targets/gpu/include/migraphx/gpu/context.hpp
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/context.hpp:162]: The workspace should not be stored in the context. Instead during lowering we can add a allocate instruction and then pass it as an argument to the operator. This is the way we handle the workspace in miopen.

File: src/targets/gpu/include/migraphx/gpu/context.hpp
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/context.hpp:150]: The preference pointer should be stored in the operator and created either during compile_blas or finalize.

File: src/targets/gpu/hip_gemm_impl.cpp
Comment (pfultz2) [inline src/targets/gpu/hip_gemm_impl.cpp:263]: Use managed pointers.

File: src/targets/gpu/include/migraphx/gpu/context.hpp
Comment (pfultz2) [reply src/targets/gpu/include/migraphx/gpu/context.hpp:150]: Actually it should just be created in the solution class. anyways.

File: src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:30]: Create a constant variable, dont use a macro.

File: src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:32]: Create a hipblaslt_invoke function to call the API and check the error code.

**Comment by pfultz2** [comment]
You need to enable hipblaslt on the tidy builds as well.

File: src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp
Comment (ahsan-ca) [reply src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp]: Done.

File: src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp
Comment (ahsan-ca) [reply src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp]: Done.

File: src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp
Comment (ahsan-ca) [reply src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp]: Done.

File: src/targets/gpu/compile_blas.cpp
Comment (ahsan-ca) [reply src/targets/gpu/compile_blas.cpp]: Removed `compile_blas`.

File: src/targets/gpu/include/migraphx/gpu/context.hpp
Comment (ahsan-ca) [reply src/targets/gpu/include/migraphx/gpu/context.hpp]: Done.

File: src/targets/gpu/include/migraphx/gpu/context.hpp
Comment (ahsan-ca) [reply src/targets/gpu/include/migraphx/gpu/context.hpp:151]: Working on this...

File: src/targets/gpu/hip_gemm_impl.cpp
Comment (ahsan-ca) [reply src/targets/gpu/hip_gemm_impl.cpp]: I was running into issues (test case failures) with using managed pointers for this. Need to look deeper into this. Any recommendations(example) on how to use managed pointers for this are welcome :).

File: src/targets/gpu/include/migraphx/gpu/context.hpp
Comment (ahsan-ca) [reply src/targets/gpu/include/migraphx/gpu/context.hpp]: Made the changes to add `allocate_gpu` in lowering.

File: src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:33]: remove these ifdefs

**Comment by causten** [comment]
testing with this branch is working.  Just need some approvals.  @TedThemistokleous @CharlieL7

**Review by TedThemistokleous — APPROVED** [review]

**Comment by TedThemistokleous** [comment]
@ahsan-ca make sure we don't also need hipblas, not just hipblaslt and -dev/devel equivalent like we had in the SLES docker case if we're still missing deps.

File: src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp
Comment (ahsan-ca) [reply src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp]: Done. I have also addressed the other comments that we discussed, in person, on Monday. Can you please also take a look at the last few commits to see if the comments have been addressed correctly? Thanks a lot for the help!

File: src/targets/gpu/include/migraphx/gpu/context.hpp
Comment (ahsan-ca) [reply src/targets/gpu/include/migraphx/gpu/context.hpp]: So the change in lowering for `allocate_gpu` would not work. I have removed the change and I will discuss this with Paul when he is back to see what we should do here.

**Comment by ahsan-ca** [comment]
> > testing with this branch is working. Just need some approvals. @TedThemistokleous @CharlieL7
> 
> @causten I discussed the changes (made for addressing comments) with Umang yesterday, I need to update some things in these changes and fix some CI failures too. I will update here once done. Thanks!

I have made all the required changes and updated the PR with latest develop. This is good to go after CI run is green. Thanks! @CharlieL7 its ready for review. Thanks!

**Review by CharlieL7 — APPROVED** [review]

File: src/targets/gpu/include/migraphx/gpu/context.hpp
Comment (pfultz2) [reply src/targets/gpu/include/migraphx/gpu/context.hpp]: What the issue that you ran into?

File: src/targets/gpu/lowering.cpp
Comment (pfultz2) [inline src/targets/gpu/lowering.cpp:253]: You need to do another `insert_allocation` for the workspace as well.

**Review by pfultz2 — CHANGES_REQUESTED** [review]

File: src/targets/gpu/include/migraphx/gpu/context.hpp
Comment (pfultz2) [reply src/targets/gpu/include/migraphx/gpu/context.hpp]: The workspace should be added before the output parameter.

**Comment by ahsan-ca** [comment]
> You need to enable hipblaslt on the tidy builds as well.

@pfultz2 I was looking at this and the cmake argument for hipblaslt is turned on by default `option(MIGRAPHX_USE_HIPBLASLT "Enable MIGraphX to use hipBLASLt" ON)`. 
`MIGRAPHX_ENABLE_HIPBLASLT_GEMM` is an runtime environment variable that controls turning hipblaslt on/off for code.
In this case, we probably don't need to add anything to the `ci.yaml` file for CI.
Is my understanding correct?

File: src/targets/gpu/lowering.cpp
Comment (ahsan-ca) [reply src/targets/gpu/lowering.cpp:259]: Made changes for allocating workspace in lowering.

File: src/targets/gpu/include/migraphx/gpu/context.hpp
Comment (ahsan-ca) [reply src/targets/gpu/include/migraphx/gpu/context.hpp]: Done.

File: hip-clang.docker
Comment (pfultz2) [inline hip-clang.docker:34]: This should be installed in the install_preqs.sh file, not in the dockerfile.

File: src/targets/gpu/CMakeLists.txt
Comment (pfultz2) [inline src/targets/gpu/CMakeLists.txt:55]: Why are we looking for hipblas? We dont use it.

Comment (ahsan-ca) [reply src/targets/gpu/CMakeLists.txt]: Removed.

File: hip-clang.docker
Comment (ahsan-ca) [reply hip-clang.docker]: Moved to install_prereqs file.

File: tools/docker/sles.docker
Comment (pfultz2) [inline tools/docker/sles.docker:30]: These should also be added to the install_preqs.

File: tools/install_prereqs.sh
Comment (pfultz2) [inline tools/install_prereqs.sh:56]: dont we need to install the dev package?

File: tools/docker/sles.docker
Comment (ahsan-ca) [reply tools/docker/sles.docker]: Done.

File: tools/install_prereqs.sh
Comment (ahsan-ca) [reply tools/install_prereqs.sh]: Updated.

File: src/targets/gpu/CMakeLists.txt
Comment (ahsan-ca) [reply src/targets/gpu/CMakeLists.txt]: Seems like hipblaslt uses hipblas. I saw sles failure after removing hipblas:
```
/opt/rocm-6.1.3/include/hipblaslt/hipblaslt.h:49:10: fatal error: 'hipblas/hipblas.h' file not found
   49 | #include <hipblas/hipblas.h>
      |          ^~~~~~~~~~~~~~~~~~~
```

Comment (ahsan-ca) [reply src/targets/gpu/CMakeLists.txt:55]: This seems to be causing this error too:
```
File "/src/AMDMIGraphX/tools/accuracy/accuracy_checker.py", line 26, in
import migraphx
ImportError: /src/AMDMIGraphX/build/lib/libmigraphx_gpu.so.2011000: undefined symbol: hipblasSetStream
```

File: Dockerfile
Comment (ahsan-ca) [inline Dockerfile:59]: Remove this and libgfortran5.

File: src/targets/gpu/CMakeLists.txt
Comment (ahsan-ca) [reply src/targets/gpu/CMakeLists.txt:55]: Ahsan to put a comment to add that we are doing this to workaround the broken hipblaslt package.
To Do:
1. Verify with hipBLASLt team if this is not fixed and open an issue to fix the hipblaslt package:
   a. Dev package needs to be fixed to inclue hipblas header files.
   b. Automatically add hipblas in cmake.

Comment (ahsan-ca) [reply src/targets/gpu/CMakeLists.txt:56]: Added comment to explain its a workaround.

File: Dockerfile
Comment (ahsan-ca) [reply Dockerfile]: Done.

File: src/targets/gpu/CMakeLists.txt
Comment (ahsan-ca) [reply src/targets/gpu/CMakeLists.txt:56]: Chris checked and the answer to the broken issue is:
```
In 6.2, hipBLAS is a dependency of hipBLASLt, so you should be able to use them together.
In 6.3, the dependencies got change around a bit and you will need to also install a package
called hipBLAS-common.
When you are adding your code now, please don’t assume that just because you have
hipBLASlt installed that means you will have hipBLAS, since this assumption will go away in 6.3.
If you need hipBLAS, install it explicitly and #include it explicitly.
```
This explains how the code in develop that we were looking at would work. They have a new package hipBLAS-common.
```
#ifndef LEGACY_HIPBLAS_DIRECT
#include <hipblas-common/hipblas-common.h>
#else
#include <hipblas/hipblas.h>
#endif
```
@pfultz2 In light of above, do we want to keep hipblas for now since CI is at 6.1. Once we update the CI to 6.3, we can remove hipblas requirement and add `hipBLAS-common` requirement. Does this seem a reasonable approach for now?

Comment (ahsan-ca) [reply src/targets/gpu/CMakeLists.txt:56]: I missed the step of updating CI to 6.2 before we update it to 6.3 :). I imagine that 6.2 should work fine though (with current workaround), since the changes for hipblaslt side went in, in 6.3.

Comment (pfultz2) [reply src/targets/gpu/CMakeLists.txt:56]: Yea we need a comment for the workaround since there package is broken in 6.1(and possibly 6.2 as well).

If hipBLAS-common package is needed in 6.3, then they need to make sure its included by default. We are not using hipblas at all.

File: src/targets/gpu/hipblaslt.cpp
Comment (pfultz2) [inline src/targets/gpu/hipblaslt.cpp:43]: This function can be removed, we already pass the stream when we call hipblasLtMatmul.

Comment (ahsan-ca) [reply src/targets/gpu/hipblaslt.cpp:43]: Sounds good. I will do that.

File: test/gpu/gemm_tune.cpp
Comment (bpickrel) [inline test/gpu/gemm_tune.cpp:134]: After using `target_compile_definitions` to set `MIGRAPHX_USE_HIPBLASLT` in the makefile, it appears to me that it will always be defined and set to either 0 or 1 at this point, which would make this `#if` always true.  Can you confirm that the code really doesn't get compiled when we don't want it to?

File: src/targets/gpu/hipblaslt.cpp
Comment (ahsan-ca) [reply src/targets/gpu/hipblaslt.cpp]: I have removed the function.

File: test/gpu/gemm_tune.cpp
Comment (ahsan-ca) [reply test/gpu/gemm_tune.cpp:134]: Thanks for catching this. I will update this.

Comment (ahsan-ca) [reply test/gpu/gemm_tune.cpp]: I have updated it to check for the value instead of definition.

**Comment by ahsan-ca** [comment]
The Windows build is passing now.
The PR is good for review, all comments, so far, have been addressed.

File: test/gpu/gemm_tune.cpp
Comment (bpickrel) [reply test/gpu/gemm_tune.cpp]: I wish we had a convention for preprocessor definitions that would make this mistake less likely, but the fix looks good.

**Review by bpickrel — APPROVED** [review]

File: src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:86]: These functions should be in the migraphx::gpu namespace.

File: src/targets/gpu/include/migraphx/gpu/hip_gemm_impl.hpp
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/hip_gemm_impl.hpp:36]: These shoudl be in the migraphx::gpu namespace as well.

File: src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:38]: Function names should use snake_case. I dont know why tidy isn't complaining.

File: src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp
Comment (ahsan-ca) [reply src/targets/gpu/include/migraphx/gpu/hipblaslt.hpp:90]: Done.

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**
